api/src/main/java/org/openmrs/module/reporting/config/ReportLoader.java

```diff
@@ -0,0 +1,251 @@
+            SerializedObject invalidSerializedObject = Context.getRegisteredComponents(SerializedObjectDAO.class).get(0).getSerializedObjectByUuid(reportDefinition.getUuid());
+            if (invalidSerializedObject != null) {
+                reportDefinition.setId(invalidSerializedObject.getId());
+            }
+        }
+
+        Context.getService(ReportDefinitionService.class).saveDefinition(reportDefinition);
+    }
+
+    public static void saveReportDesigns(ReportDefinition reportDefinition, List<ReportDesign> reportDesigns) {
+        // purging a ReportDesign doesn't trigger any extra logic, so we can just purge-and-recreate here
+        List<ReportDesign> existingDesigns = Context.getService(ReportService.class).getReportDesigns(reportDefinition, null, true);
+        if (existingDesigns.size() > 0) {
+            log.debug("Deleting " + existingDesigns.size() + " old designs for " + reportDefinition.getName());
+            for (ReportDesign design : existingDesigns) {
+                Context.getService(ReportService.class).purgeReportDesign(design);
+            }
+        }
+
+        for (ReportDesign reportDesign : reportDesigns) {
+            Context.getService(ReportService.class).saveReportDesign(reportDesign);
+        }
+
+    }
+
+    public static  ReportDefinition constructReportDefinition(ReportDescriptor reportDescriptor) {
+
+        ReportDefinition rd = new ReportDefinition();
+        // TODO do we need to handle translations here?
+        rd.setName(reportDescriptor.getName());
+        rd.setDescription(reportDescriptor.getDescription());
+        rd.setUuid(reportDescriptor.getUuid());
+
+        List<Parameter> parameters = constructParameters(reportDescriptor.getParameters());
+        Map<String, Object> mappings = constructMappings(parameters);
+        rd.setParameters(parameters);
+
+        if (reportDescriptor.getDatasets() != null) {
+            for (DataSetDescriptor dataSetDescriptor : reportDescriptor.getDatasets()) {
+                DataSetDefinition dsd = constructDataSetDefinition(dataSetDescriptor, parameters);
+                // TODO do we need to handle DSD name and description?
+                if (dsd != null) {
+                    rd.addDataSetDefinition(dataSetDescriptor.getKey(), dsd, mappings);
+                }
+            }
+        }
+        return rd;
+    }
+
+    public static List<Parameter> constructParameters(List<ParameterDescriptor> parameterDescriptors) {
+        List<Parameter> parameters = new ArrayList<Parameter>();
+        if (parameterDescriptors != null) {
+            for (ParameterDescriptor parameterDescriptor : parameterDescriptors) {
+                try {
+                    // TODO how else might we want to configure parameters? ie other "types" besides classes?
+                    parameters.add(new Parameter(parameterDescriptor.getKey(), parameterDescriptor.getLabel(), Class.forName(parameterDescriptor.getType())));
```

Comment:
Not sure about the Class.forName() here.  Does this do what you/we expect?  In your samplereport.yml file above, you have type: "date".  Will this be handled as expected?

The other thing is that in OpenMRS, we should use Context.loadClass(String classname) as this will ensure that the OpenmrsClassLoader is used, which is what we want.

What I would do here is support fully-specified class names using Context.loadClass, with a few hard-coded shortcuts for "date" and "location" that simply map these to java.util.Date and org.openmrs.Location before handling.
 probably support all fully-specified class names (you should use Context.get

Reply:
Yeah, "date" won't work... I switched some of the test data but missed that.

Anyway, your suggestions make sense... are there any other hardcoded ones you think we will want to support beyond Location and Date?

Reply:
I’d just stick with those 2 for now. Can always add more in